src/BreakpointViewer.cpp

```diff
@@ -0,0 +1,621 @@
+	}
+	return true;
+}
+
+bool BreakpointViewer::processLocationField(int index, BreakpointViewer::Type type,
+		const QString& field, int& begin, int& end)
+{
+	if (type == BreakpointViewer::BREAKPOINT) {
+		int value;
+		if ((value = stringToValue(field.simplified())) != -1) {
+			begin = 0xFFFF & value;
+			end = -1;
+			return true;
+		} else if (index != -1) {
+			begin = breakpoints->getAddress(index);
+			end = -1;
+			return true;
+		}
+		return false;
+	}
+	QStringList s = field.simplified().split(":", Qt::SplitBehaviorFlags::SkipEmptyParts);
+	int value;
+	if (!s.length() || (value = stringToValue(s[0])) == -1) {
+		if (index != -1) {
+			begin = breakpoints->getAddress(index);
+			end = breakpoints->getAttribute(index, "regionEnd").toInt();
+			return true;
+		}
+		return false;
+	}
+	begin = value;
+	if (s.length() > 2 || (s.length() == 2 && (value = stringToValue(s[1])) == -1)) {
+		if (index != -1) {
+			begin = breakpoints->getAddress(index);
+			end = breakpoints->getAttribute(index, "regionEnd").toInt();
+			return true;
+		}
+		return false;
+	}
+	end = value;
+	return true;
+}
+
+void BreakpointViewer::changeTableItem(Type type, QTableWidgetItem* item)
+{
+	auto table = selector[type];
+
+	// trying to modify a bp instead of creating a new one
+	bool createBp = false;
+
+	// check if breakpoint is enabled
+	int row = table->row(item);
+	int index = items.indexOf({type, row});
+	QTableWidgetItem* enabledItem = table->item(row, ENABLED);
+	bool enabled = enabledItem->checkState() == Qt::Checked;
+
+	switch (table->column(item)) {
+		case ENABLED:
+			createBp = enabled;
+			break;
+		case WP_TYPE:
+			assert(table == wpTableWidget);
+			break;
+		case LOCATION: {
+			int adrLen;
+
+			if (type == WATCHPOINT) {
+				auto model = table->model();
+				auto combo = (QComboBox*) table->indexWidget(model->index(row, WP_TYPE));
+				Breakpoints::Type wtype = static_cast<Breakpoints::Type>(combo->currentIndex() + 1);
+				adrLen = (wtype == Breakpoints::WATCHPOINT_IOREAD || wtype == Breakpoints::WATCHPOINT_IOWRITE)
+					? 2 : 4;
+			} else {
+				adrLen = 4;
+			}
+			int begin, end;
+
+			if (processLocationField(index, type, item->text(), begin, end)) {
+				userMode = false;
+				item->setText(QString("%1%2%3").arg(hexValue(begin, adrLen))
+					.arg(end == begin || end == -1 ? "" : ":")
+					.arg(end == begin || end == -1 ? "" : hexValue(end, adrLen)));
+				userMode = true;
+			} else {
+				userMode = false;
+				enabled = false;
+				item->setText("");
+				enabledItem->setCheckState(Qt::Unchecked);
+				userMode = true;
+			}
+			if (!enabled) return;
+			break;
+		}
+		case SLOT: {
+			qint8 ps, ss;
+			if (processSlotField(index, item->text(), ps, ss)) {
+				userMode = false;
+				item->setText(QString("%1/%2")
+					.arg(ps == -1 ? QChar('X') : QChar('0' + ps))
+					.arg(ss == -1 ? QChar('X') : QChar('0' + ss)));
+				userMode = true;
+			} else {
+				userMode = false;
+				item->setText("X/X");
+				userMode = true;
```

Comment:
In many locations you have something like this:
* Temporary set `userMode` to `false`.
* Then restore to `true`.

This was unexpected to me. I assume it's needed to break some update loops?

If this is a frequent pattern, you can consider a helper like this (requires c++17):
```
       auto sa = scoped_assign(userMode, false);
       item->setText("X/X");
       // at this point 'sa' goes out of scope, and 'userMode' is restored to it's previous value
```
Advantage is:
 * You cannot forget to restore `userMode`. Also not if you leave this scope via an early return or via an exception.
 * The value is restored to the previous value (which is not necessarily the value `true`). This might be useful if in the future you have more nested calls.

Reply:
> This was unexpected to me. I assume it's needed to break some update loops?

Yes. Qt doesn't distinguish user generated events from programmed events. They both trigger widget-changed events that call slots attached to them. But the program needs to sanitize input, so it will update the widgets, but we don't want events being triggered again (which will probably loop forever anyway since event-slot don't use the stack).

> If this is a frequent pattern, you can consider a helper like this (requires c++17):

That's a great idea. But I couldn't find that function (class?), so I created my own:
```
template <class T>
class scoped_assign
{
private:
	T& variable;
	T oldValue;
public:
	scoped_assign(T& var, T value) : variable(var) {
		oldValue = var;
	}

	~scoped_assign() {
		variable = oldValue;
	}
};
```

Reply:
Yes, I meant something like the ScopedAssign helper we have in openmsx:
https://github.com/openMSX/openMSX/blob/master/src/utils/ScopedAssign.hh
Detail: in your implementation you forgot to assign `value` to `var`.

To use it prior to c++17 you'd have to write something like:
```scoped_assign<bool> sa(userMode, false);```
Starting from c++17 you can simplify that to:
```scoped_assign sa(userMode, false);```

Reply:
Right, I will import it from OpenMSX. But we should do this in the constructor:
```
	[[nodiscard]] ScopedAssign(T& var_, T newValue)
```